tests/unit/ViolationUtilsTest.js

```diff
@@ -0,0 +1,505 @@
+                        userMessage: '',
+                    },
+                    ...transactionViolations,
+                ]),
+            );
+        });
+    });
+
+    describe('missing or invalid data', () => {
+        it('should handle empty transaction', () => {
+            // setup test conditions
+            transaction = {};
+            transactionViolations = [];
+            policyRequiresCategories = undefined;
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual([]);
+        });
+
+        it('should handle nonexistent category in policy', () => {
+            // setup test conditions
+            transaction.category = 'Nonexistent';
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(expect.arrayContaining([{name: 'categoryOutOfPolicy', type: 'violation', userMessage: ''}]));
+        });
+
+        it('should handle nonexistent tag in policy', () => {
+            // setup test conditions
+            transaction.tag = 'Nonexistent';
+            policyRequiresTags = true;
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(expect.arrayContaining([{name: 'tagOutOfPolicy', type: 'violation', userMessage: ''}]));
+        });
+
+        it('should handle invalid policy settings', () => {
+            // setup test conditions
+            policyRequiresTags = true;
+            policyTags = {};
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(
+                expect.arrayContaining([
+                    {
+                        name: 'tagOutOfPolicy',
+                        type: 'violation',
+                        userMessage: '',
+                    },
+                ]),
+            );
+        });
+
+        it('should handle case sensitivity', () => {
+            // setup test conditions
+            transaction.category = 'food';
+            // run test
+            const result = execute();
+            // verify results
+            expect(result.value).toEqual(
+                expect.arrayContaining([
+                    {
+                        name: 'categoryOutOfPolicy',
+                        type: 'violation',
+                        userMessage: '',
+                    },
+                ]),
+            );
+        });
+
+        it('should handle null or undefined values', () => {
+            // setup test conditions
+            transaction = null;
+            policyRequiresTags = null;
+            policyTags = null;
+            policyRequiresCategories = null;
+            policyCategories = null;
+            // run test
+            expect(() => execute()).toThrow();
+        });
+
+        it('should handle invalid data types', () => {
```

Comment:
We shouldn't test this either imo. Our code shouldn't allow us to set transaction.category to something other than a string